File: framework/ruby/src/soda/leetcode/list.rb
module Soda end

module Soda::Leetcode

class ListNode
  attr_accessor :val, :next
  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end
end

class ListFactory
  def self.create(data)
    head = ListNode.new
    tail = head
    data.each { |val|
      node = ListNode.new(val)
      tail.next = node
      tail = node
    }
    head.next
  end

  def self.dump(head)
    list = []
    while head
      list << head.val
      head = head.next
    end
    list
  end
end

end  # module Soda::Leetcode

File: framework/js/src/soda/leetcode/list.js
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

class ListFactory {
    static create(data) {
        let head = new ListNode();
        let tail = head;
        data.forEach(val => {
            let node = new ListNode(val);
            tail.next = node;
            tail = node;
        });
        return head.next;
    }

    static dump(head) {
        let list = [];
        while (head) {
            list.push(head.val);
            head = head.next;
        }
        return list;
    }
}

exports.ListNode = ListNode;
exports.ListFactory = ListFactory;
